sdks/python/apache_beam/runners/interactive/caching/streaming_cache.py

```diff
@@ -0,0 +1,171 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+from __future__ import absolute_import
+
+from apache_beam.portability.api.beam_runner_api_pb2 import TestStreamPayload
+from apache_beam.utils import timestamp
+from apache_beam.utils.timestamp import Timestamp
+
+
+class StreamingCache(object):
+  """Abstraction that holds the logic for reading and writing to cache.
+  """
+  def __init__(self, readers):
+    self._readers = readers
+
+  class Reader(object):
+    """Abstraction that reads from PCollection readers.
+
+    This class is an Abstraction layer over multiple PCollection readers to be
+    used for supplying a TestStream service with events.
+
+    This class is also responsible for holding the state of the clock, injecting
+    clock advancement events, and watermark advancement events.
+    """
+    def __init__(self, readers):
+      # This timestamp is used as the monotonic clock to order events in the
+      # replay.
+      self._monotonic_clock = timestamp.Timestamp.of(0)
+
+      # The PCollection cache readers.
+      self._readers = {}
+
+      # The file headers that are metadata for that particular PCollection.
+      # The header allows for metadata about an entire stream, so that the data
+      # isn't copied per record.
+      self._headers = {r.header().tag : r.header() for r in readers}
+      self._readers = {r.header().tag : r.read() for r in readers}
+
+      # The watermarks per tag. Useful for introspection in the stream.
+      self._watermarks = {tag: timestamp.MIN_TIMESTAMP for tag in self._headers}
+
+      # The most recently read timestamp per tag.
+      self._stream_times = {tag: timestamp.MIN_TIMESTAMP
+                            for tag in self._headers}
+
+    def _test_stream_events_before_target(self, target_timestamp):
+      """Reads the next iteration of elements from each stream.
+
+      Retrieves an element from each stream iff the most recently read timestamp
+      from that stream is less than the target_timestamp. Since the amount of
+      events may not fit into memory, this StreamingCache needs to read each
+      element one at a time.
+      """
+      records = []
+      for tag, r in self._readers.items():
+        # The target_timestamp is the maximum timestamp that was read from the
+        # stream. Some readers may have elements that are less than this. Thus,
+        # we skip all readers that already have elements that are at this
+        # timestamp so that we don't read everything into memory.
+        if self._stream_times[tag] >= target_timestamp:
+          continue
+        try:
+          record = next(r)
+          records.append((tag, record))
+          self._stream_times[tag] = Timestamp.from_proto(record.processing_time)
+        except StopIteration:
+          pass
+      return records
+
+    def _merge_sort(self, previous_events, new_events):
```

Comment:
By merge sort, I meant basically https://guava.dev/releases/20.0/api/docs/com/google/common/collect/Iterables.html#mergeSorted-java.lang.Iterable-java.util.Comparator-

In short, one maintains a heap of readers sorted by the timestamp of their (peeked at) next element. Then

```
while not reader_heap.empty():
  first_reader = reader_heap.pop()
  try:
    yield first_reader.read()
  except StopIteration:
    pass
  else:
    reader_heap.push(first_reader)
```

This should be equivalent, so you can keep things as is if you want.

Reply:
Ack, I think that would be a good refactor for a follow-up PR.